File: python_data-structures_algorithms/side_trail/maximum_subarray_elements.py


def maximum_subarray_elements(arr):
    max_sum = arr[0]
    curr_sum = arr[0]
    start_index = 0
    end_index = 0
    temp_start_index = 0

    for i in range(1, len(arr)):
        print(f"max_sum: {max_sum}")
        if curr_sum >= 0:
            curr_sum += arr[i]
        else:
            curr_sum = arr[i]

        if max_sum < curr_sum:
            max_sum = curr_sum
            end_index = i + 1
            if temp_start_index != 0:
                start_index = temp_start_index

        if curr_sum < 0 :
            temp_start_index = i + 1
        print(start_index, end_index)
    max_elements = arr[start_index: end_index]
    return max_elements







if __name__ == "__main__":
    arr = [5, 6, -3, 7, -13, 8, -2, 5, -6, 7, -11, 3, 10, -10, -6, -10, 7, 2]

    elements = maximum_subarray_elements(arr)

    print(elements)

    

File: java_17/challenge_questions/arrays_and_strings/MaximumSubarrayElements.java
package java_17.challenge_questions.arrays_and_strings;

import java.util.Arrays;

public class MaximumSubarrayElements {

    public static void main(String[] args) {

        int[] arr = {5, 6, -3, 7, -13, 8, -2, 5, -6, 7, -11, 3, 10, -10, -6, -10, 7, 10};

        System.out.println(Arrays.toString(maximumSubarrayElements(arr)));
        
        int[] arr2 = {-1, -1, 0, 0, -1};
        
        System.out.println(Arrays.toString(maximumSubarrayElements(arr2)));
        
    }

    public static int[] maximumSubarrayElements(int[] arr) {
        int maxSum = arr[0];
        int currSum = arr[0];
        int startIndex = 0;
        int endIndex = 0;
        int tempStartIndex = 0;

        for (int i = 1; i < arr.length; i++) {
            if (currSum >= 0) {
                currSum += arr[i];
            } else {
                currSum = arr[i];
            }
            if (currSum >= maxSum) {
                maxSum = currSum;
                endIndex = i + 1;
                if (tempStartIndex != 0) {
                    startIndex = tempStartIndex;
                }
            }

            if (currSum < 0) {
                tempStartIndex = i + 1;
            }
        }   
        int[] maxElements = Arrays.copyOfRange(arr, startIndex, endIndex);
        return maxElements;
    }
}